### DISCUSSION TITLE
[python] inline quotation to show the value of a variable

### DISCUSSION THREAD

**Discussion by sophia-hye** [quarto-dev/quarto-cli#5638] [Q&A] [answered]
Hello, there 👋
Before I ask a question, the Quarto is very useful to me so I wanna say "thanks".

❓ 
When I want to show the result (e.g., df.shape, len([1,2,3], ...) or the value of a variable when I used it in some python code cell, I can't insert it between words in markdown.

For example,
If my python code like below,
```{python}
mylist = [1,2,3]
...
user_names = df[(df.age < 10)].name.tolist()
```
My expectation is this:

> ... the length of the `mylist` is 3. ...
> ... This is the user name under 10 years old. : ['AAA', 'BBB', 'CCC', 'DDD'] ...

Is it impossible in this version? or Is there some way to do this?

**Comment by baptiste** [comment]
I believe this isn't possible with the Jupyter engine, because the Jupyter project hasn't reached consensus on how to implement this in markdown cells. As far as I can tell the numerous online discussions never went past early proof-of-principles and further debates.

If you're using the knitr engine, you can pass data between Python cells and R, and use the `` `r myvalue` `` syntax.


------

I wonder if Quarto could work around the jupyter hold-up, with the following strategy:

- use a custom span syntax for such inline code cells, e.g. ```[`len(mylist)`]{.python}``` 
- such content gets picked up early enough (custom pre-processing step needed, presumably) and gets converted into a kind of _shadow_ code cell (or a real one, if conversion to jupyter notebook is the goal), that gets evaluated alongside all other cells in the standard way
- the output of these shadow cells is flagged as needing to be folded back inline, instead of appearing as a standard output environment
- these values are just new nodes in the AST at this point

**Reply by sophia-hye** [reply]
Thank you, @baptiste 
Can I get some example how to include excutable expressions within markdown using `r ` expression?

I try to include the value of a variable within markdown like  `r myvalue`  using knitr engine as you said, but the qmd file can't be rendered and error occured like this: 

> Error in eval(parse_only(code), envir = envir) : 
> object `python_value` not found 
> Calls: .main ... inline_exec -> hook_eval -> withVisible -> eval -> eval
> Execution halted

`python_value` is the value of the python
```{python}
python_value = df.shape[0]
```
 and the markdown is this
> Test to include executable expressions within markdown: `r python_value`.

**Reply by baptiste** [reply, marked as answer]
You need to use reticulate to pass values from Python to R,

````
---
title: "Untitled"
format: html
engine: knitr
---

```{r, setup}
library(reticulate)
```

Create a variable `x` in the Python session:

```{python}
x = [1, 2, 3]
```

Access the Python variable `x` in an R code chunk:

```{r}
py$x
```

or in `r py$x` the text.

````

**Reply by sophia-hye** [reply]
Thank you @baptiste 😄

**Comment by cscheid** [comment]
> I believe this isn't possible with the Jupyter engine, because the Jupyter project hasn't reached consensus on how to implement this in markdown cells. As far as I can tell the numerous online discussions never went past early proof-of-principles and further debates.

This is an accurate statement. With that said, we do think we now have a workable proposal for a limited, language-agnostic solution to this problem. I don't expect it to make to 1.4, but hopefully in 1.5.

**Comment by baptiste** [comment]
As a side-note, `ojs` provides a rather lightweight workaround with its markdown interpolation. It still requires wrapping the text in a code cell with a `` md` ` `` wrapper, but it's not too obtrusive. The nonlocal evaluation is something to keep in mind though., but the `ojs_define()` step makes the intention clear.

````
---
title: "Untitled"
format: html
---

```{r, echo=-c(4:5)}
amount <- 800e3
interest <- 5
duration <- 20
ojs_define(interest)
ojs_define(duration)
```

```{ojs}
//| echo: false
md`A mortgage for a house purchased for **${amount}** at an interest rate of **${interest}%** over **${duration} years** costs a total of **${amount * (1+interest/100) ** duration}**`
```


```{r, echo=-2}
amount <- 8
ojs_define(amount)
```

````

**Reply by cscheid** [reply]
(it won't surprise you that our solution is indeed a minimal language-agnostic generalization of ojs_define..)
